File: go/src/inversions/inversions.go
/*
	Counting inversions
*/
package inversions

import (
	"math"
)

func CountInversions(data []int) int {
	count := 0
	mergeSort(data, &count)
	return count
}

func mergeSort(data []int, count *int) []int {
	n := len(data)
	if n < 2 {
		return data
	}
	mid := getMedian(n)
	left := mergeSort(data[:mid], count)
	right := mergeSort(data[mid:], count)
	return merge(left, right, count)
}

func merge(left, right []int, count *int) []int {
	n := len(left) + len(right)
	result := make([]int, n)
	i, j, k := 0, 0, 0
	for i < len(left) && j < len(right) {
		if left[i] < right[j] {
			result[k] = left[i]
			i++
		} else {
			result[k] = right[j]
			*count += len(left[i:])
			j++
		}
		k++
	}
	for i < len(left) {
		result[k] = left[i]
		i++
		k++
	}
	for j < len(right) {
		result[k] = right[j]
		j++
		k++
	}
	return result
}

func getMedian(n int) int {
	return int(math.Floor(float64(n) / float64(2)))
}

File: python/inversions/inversions.py
#!/usr/bin/env python
#
# Counting inversions
#
import sys
import math

def count_inversions(a):
        data = { 'count': 0 }
        merge_sort(a, data)
        return data['count']

def merge_sort(a, data): 
    if len(a) < 2:
        return a
    else:
        mid = int(math.floor(len(a) / 2.0))
        merged = []
        l = merge_sort(a[:mid], data)
        r = merge_sort(a[mid:], data)
        return merge(l, r, data)

def merge(left, right, data): 
    result = []

    i = j = 0
    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            result.append(left[i])
            i = i + 1
        else:
            result.append(right[j])
            j = j + 1
            # count the inversions
            data['count'] = data['count'] + len(left[i:])

    while i < len(left):
        result.append(left[i])
        i = i + 1

    while j < len(right):
        result.append(right[j])
        j = j + 1

    return result 

def num(s):
    try:
        return int(s)
    except ValueError:
        return float(s)



